File: PersonalizedAndroid/src/com/example/personalizedandroid/MainActivity.java
package com.example.personalizedandroid;

import android.app.Activity;
import android.app.ActivityManager;
import android.app.IActivityManager;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.os.RemoteException;
import android.os.UserManager;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.TextView;
import android.content.pm.UserInfo;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends Activity {
    private static final String TAG = "PersonalizedAndroid";
    private UserManager mUserManager;
    private IActivityManager mActivityManager;
    private TextView mUserListTextView;
    private EditText mUserNameInput;
    private Spinner mUserSpinner;
    private ArrayAdapter<String> mUserAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mUserListTextView = findViewById(R.id.user_list);
        mUserNameInput = findViewById(R.id.user_name_input);
        mUserSpinner = findViewById(R.id.user_spinner);
        Button createUserButton = findViewById(R.id.create_user_button);
        Button deleteUserButton = findViewById(R.id.delete_user_button);

        // Initialize UserManager
        mUserManager = getSystemService(UserManager.class);

        // Initialize ActivityManager
        mActivityManager = ActivityManager.getService();

        // Check and restore intended user
        int intendedUserId = getIntent().getIntExtra(WelcomeActivity.EXTRA_USER_ID, -1);
        int currentUserId = getCurrentUserId();
        Log.d(TAG, "MainActivity: Intended user ID: " + intendedUserId + ", Current user ID: " + currentUserId);
        if (intendedUserId != -1 && intendedUserId != currentUserId && intendedUserId != 0) {
            switchUser(intendedUserId);
        }

        // Apply current user's preferences
        applyPreferences();

        // Set up user spinner
        setupUserSpinner();

        // List existing users
        listUsers();

        // Set up button to create a new user
        createUserButton.setOnClickListener(v -> {
            String name = mUserNameInput.getText().toString().trim();
            if (!TextUtils.isEmpty(name)) {
                createNewUser(name);
                mUserNameInput.setText(""); // Clear input after creation
                setupUserSpinner(); // Refresh spinner
            } else {
                mUserListTextView.setText("Please enter a user name");
            }
        });

        // Set up button to delete selected user
        deleteUserButton.setOnClickListener(v -> {
            int selectedPosition = mUserSpinner.getSelectedItemPosition();
            if (selectedPosition != AdapterView.INVALID_POSITION) {
                List<UserInfo> users = mUserManager.getUsers();
                if (selectedPosition < users.size()) {
                    UserInfo selectedUser = users.get(selectedPosition);
                    deleteUser(selectedUser.id);
                    setupUserSpinner(); // Refresh spinner
                }
            }
        });
    }

    private void setupUserSpinner() {
        if (mUserManager != null) {
            List<UserInfo> users = mUserManager.getUsers();
            List<String> userNames = new ArrayList<>();
            for (UserInfo user : users) {
                userNames.add("ID: " + user.id + ", Name: " + user.name);
            }
            mUserAdapter = new ArrayAdapter<>(this, android.R.layout.simple_spinner_item, userNames);
            mUserAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
            mUserSpinner.setAdapter(mUserAdapter);

            // Switch user on selection
            mUserSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
                @Override
                public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                    List<UserInfo> users = mUserManager.getUsers();
                    if (position < users.size()) {
                        UserInfo selectedUser = users.get(position);
                        if (selectedUser.id != getCurrentUserId()) {
                            switchUser(selectedUser.id);
                        }
                    }
                }

                @Override
                public void onNothingSelected(AdapterView<?> parent) {}
            });
        }
    }

    private void listUsers() {
        if (mUserManager != null) {
            List<UserInfo> users = mUserManager.getUsers();
            StringBuilder userText = new StringBuilder("Users:\n");
            for (UserInfo user : users) {
                userText.append("ID: ").append(user.id)
                        .append(", Name: ").append(user.name)
                        .append(", Theme: ").append(loadUserPreferences(user.id)).append("\n");
            }
            mUserListTextView.setText(userText.toString());
        } else {
            mUserListTextView.setText("UserManager not initialized");
        }
    }

    private void createNewUser(String name) {
        if (mUserManager != null) {
            try {
                UserInfo newUser = mUserManager.createUser(name, UserManager.USER_TYPE_FULL_SECONDARY, 0);
                if (newUser != null) {
                    Log.d(TAG, "User created: " + newUser.name);
                    mUserListTextView.setText("User created: " + newUser.name);
                    saveUserPreferences(newUser.id, "dark"); // Default theme
                    listUsers();
                } else {
                    Log.e(TAG, "Failed to create user: null result");
                    mUserListTextView.setText("Failed to create user: null result");
                }
            } catch (Exception e) {
                Log.e(TAG, "Failed to create user: " + e.getMessage());
                mUserListTextView.setText("Failed to create user: " + e.getMessage());
            }
        } else {
            mUserListTextView.setText("UserManager not initialized");
        }
    }

    private void switchUser(int userId) {
        if (mActivityManager != null) {
            try {
                mActivityManager.switchUser(userId);
                Log.d(TAG, "Switched to user ID: " + userId);
                mUserListTextView.setText("Switched to user ID: " + userId);
                applyPreferences();
                listUsers();
            } catch (RemoteException e) {
                Log.e(TAG, "Failed to switch user: " + e.getMessage());
                mUserListTextView.setText("Failed to switch user: " + e.getMessage());
            }
        } else {
            mUserListTextView.setText("ActivityManager not initialized");
        }
    }

    private void deleteUser(int userId) {
        if (mUserManager != null) {
            if (userId == getCurrentUserId()) {
                mUserListTextView.setText("Cannot delete current user");
                return;
            }
            if (userId == 0) {
                mUserListTextView.setText("Cannot delete system user");
                return;
            }
            try {
                boolean success = mUserManager.removeUser(userId);
                if (success) {
                    Log.d(TAG, "User deleted: ID " + userId);
                    mUserListTextView.setText("User deleted: ID " + userId);
                    listUsers();
                } else {
                    Log.e(TAG, "Failed to delete user: ID " + userId);
                    mUserListTextView.setText("Failed to delete user: ID " + userId);
                }
            } catch (Exception e) {
                Log.e(TAG, "Failed to delete user: " + e.getMessage());
                mUserListTextView.setText("Failed to delete user: " + e.getMessage());
            }
        } else {
            mUserListTextView.setText("UserManager not initialized");
        }
    }

    private void saveUserPreferences(int userId, String theme) {
        SharedPreferences prefs = getSharedPreferences("user_" + userId, MODE_PRIVATE);
        prefs.edit().putString("theme", theme).apply();
    }

    private String loadUserPreferences(int userId) {
        SharedPreferences prefs = getSharedPreferences("user_" + userId, MODE_PRIVATE);
        return prefs.getString("theme", "default");
    }

    private void applyPreferences() {
        int userId = getCurrentUserId();
        String theme = loadUserPreferences(userId);
        if ("dark".equals(theme)) {
            mUserListTextView.setBackgroundColor(getResources().getColor(android.R.color.black));
            mUserListTextView.setTextColor(getResources().getColor(android.R.color.white));
        } else {
            mUserListTextView.setBackgroundColor(getResources().getColor(android.R.color.white));
            mUserListTextView.setTextColor(getResources().getColor(android.R.color.black));
        }
    }

    private int getCurrentUserId() {
        if (mUserManager != null) {
            UserInfo currentUser = mUserManager.getUserInfo(android.os.Process.myUserHandle().getIdentifier());
            if (currentUser != null) {
                Log.d(TAG, "getCurrentUserId: ID: " + currentUser.id + ", Name: " + currentUser.name);
                return currentUser.id;
            } else {
                Log.e(TAG, "getCurrentUserId: UserInfo is null");
            }
        } else {
            Log.e(TAG, "getCurrentUserId: UserManager is null");
        }
        return -1;
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }
}

File: PersonalizedAndroid/PersonalizedAndroid/src/com/example/personalizedandroid/MainActivity.kt
package com.example.personalizedandroid

import android.app.Activity
import android.app.ActivityManager
import android.app.IActivityManager
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.os.RemoteException
import android.os.UserManager
import android.text.TextUtils
import android.util.Log
import android.view.View
import android.widget.*
import android.content.pm.UserInfo

class MainActivity : Activity() {

    private lateinit var userManager: UserManager
    private var activityManager: IActivityManager? = null

    private lateinit var userListTextView: TextView
    private lateinit var userNameInput: EditText
    private lateinit var userSpinner: Spinner
    private lateinit var themeSpinner: Spinner

    private lateinit var userAdapter: ArrayAdapter<String>
    private lateinit var themeAdapter: ArrayAdapter<CharSequence>

    private var isSpinnerInitialized = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val mainLayout: LinearLayout = findViewById<LinearLayout>(R.id.main_layout)!!
userListTextView = findViewById<TextView>(R.id.user_list)!!
userNameInput = findViewById<EditText>(R.id.user_name_input)!!
userSpinner = findViewById<Spinner>(R.id.user_spinner)!!
themeSpinner = findViewById<Spinner>(R.id.theme_spinner)!!
val createUserButton: Button = findViewById<Button>(R.id.create_user_button)!!
val deleteUserButton: Button = findViewById<Button>(R.id.delete_user_button)!!


userManager = getSystemService(UserManager::class.java)!!

        activityManager = ActivityManager.getService()

        val intendedUserId = intent.getIntExtra(WelcomeActivity.EXTRA_USER_ID, -1)
        val currentUserId = getCurrentUserId()
        Log.d(TAG, "MainActivity: Intended user ID: $intendedUserId, Current user ID: $currentUserId")
        if (intendedUserId != -1 && intendedUserId != currentUserId && intendedUserId != 0) {
            switchUser(intendedUserId)
        }

applyTheme(userId, findViewById<LinearLayout>(R.id.main_layout)!!)

        applyPreferences()

        setupUserSpinner()
        setupThemeSpinner()
        listUsers()

        createUserButton.setOnClickListener {
            val name = userNameInput.text.toString().trim()
            if (!TextUtils.isEmpty(name)) {
                createNewUser(name)
                userNameInput.setText("")
                setupUserSpinner()
            } else {
                userListTextView.text = "Please enter a user name"
            }
        }

        deleteUserButton.setOnClickListener {
            val selectedPosition = userSpinner.selectedItemPosition
            if (selectedPosition != AdapterView.INVALID_POSITION) {
                val users = userManager.users
                if (selectedPosition < users.size) {
                    val selectedUser = users[selectedPosition]
                    deleteUser(selectedUser.id)
                    setupUserSpinner()
                }
            }
        }
    }

    private fun setupUserSpinner() {
        val users = userManager.users
        val userNames = ArrayList<String>()
        val currentUserId = getCurrentUserId()
        var currentUserPosition = 0

        for ((i, user) in users.withIndex()) {
            userNames.add("ID: ${user.id}, Name: ${user.name}")
            if (user.id == currentUserId) {
                currentUserPosition = i
            }
        }

        userAdapter = ArrayAdapter(this, android.R.layout.simple_spinner_item, userNames)
        userAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        userSpinner.adapter = userAdapter
        userSpinner.setSelection(currentUserPosition, false)
        isSpinnerInitialized = false

        userSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                if (!isSpinnerInitialized) {
                    isSpinnerInitialized = true
                    Log.d(TAG, "Spinner initialized, skipping switch for position: $position")
                    return
                }

                val selectedUser = users[position]
                Log.d(TAG, "Spinner selected user ID: ${selectedUser.id}, Name: ${selectedUser.name}")
                if (selectedUser.id != getCurrentUserId()) {
                    switchUser(selectedUser.id)
                }
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }
    }

    private fun setupThemeSpinner() {
        themeAdapter = ArrayAdapter.createFromResource(this, R.array.theme_options, android.R.layout.simple_spinner_item)
        themeAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        themeSpinner.adapter = themeAdapter

        val currentUserId = getCurrentUserId()
        val currentTheme = loadUserPreferences(currentUserId)

        val themePosition = when (currentTheme) {
            "solid_red" -> 1
            "solid_green" -> 2
            else -> 0
        }

        themeSpinner.setSelection(themePosition, false)

        themeSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                val theme = when (position) {
                    1 -> "solid_red"
                    2 -> "solid_green"
                    else -> "gradient_blue"
                }

                val userId = getCurrentUserId()
                saveUserPreferences(userId, theme)
                applyTheme(userId, findViewById<LinearLayout>(R.id.main_layout)!!)
                applyPreferences()
                Log.d(TAG, "Theme selected: $theme for user ID: $userId")
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }
    }

    private fun listUsers() {
        val users = userManager.users
        val userText = StringBuilder("Users:\n")
        for (user in users) {
            userText.append("ID: ${user.id}, Name: ${user.name}, Theme: ${loadUserPreferences(user.id)}\n")
        }
        userListTextView.text = userText.toString()
    }

    private fun createNewUser(name: String) {
        try {
            val newUser = userManager.createUser(name, UserManager.USER_TYPE_FULL_SECONDARY, 0)
            if (newUser != null) {
                Log.d(TAG, "User created: ${newUser.name}")
                userListTextView.text = "User created: ${newUser.name}"
                saveUserPreferences(newUser.id, "gradient_blue")
                listUsers()
            } else {
                Log.e(TAG, "Failed to create user: null result")
                userListTextView.text = "Failed to create user: null result"
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to create user: ${e.message}")
            userListTextView.text = "Failed to create user: ${e.message}"
        }
    }

    private fun switchUser(userId: Int) {
        try {
            Log.d(TAG, "Attempting to switch to user ID: $userId")
            activityManager?.switchUser(userId)
            Log.d(TAG, "Switched to user ID: $userId")
            userListTextView.text = "Switched to user ID: $userId"
            applyTheme(userId, findViewById<LinearLayout>(R.id.main_layout)!!)
            applyPreferences()
            listUsers()
        } catch (e: RemoteException) {
            Log.e(TAG, "Failed to switch user: ${e.message}")
            userListTextView.text = "Failed to switch user: ${e.message}"
        }
    }

    private fun deleteUser(userId: Int) {
        if (userId == getCurrentUserId()) {
            userListTextView.text = "Cannot delete current user"
            return
        }
        if (userId == 0) {
            userListTextView.text = "Cannot delete system user"
            return
        }
        try {
            val success = userManager.removeUser(userId)
            if (success) {
                Log.d(TAG, "User deleted: ID $userId")
                userListTextView.text = "User deleted: ID $userId"
                listUsers()
            } else {
                Log.e(TAG, "Failed to delete user: ID $userId")
                userListTextView.text = "Failed to delete user: ID $userId"
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to delete user: ${e.message}")
            userListTextView.text = "Failed to delete user: ${e.message}"
        }
    }

    private fun saveUserPreferences(userId: Int, theme: String) {
        val prefs = getSharedPreferences("user_$userId", MODE_PRIVATE)
        prefs.edit().putString("theme", theme).apply()
    }

    private fun loadUserPreferences(userId: Int): String {
        val prefs = getSharedPreferences("user_$userId", MODE_PRIVATE)
        return prefs.getString("theme", "gradient_blue") ?: "gradient_blue"
    }

    private fun applyTheme(userId: Int, layout: LinearLayout) {
        val backgroundRes = when (loadUserPreferences(userId)) {
            "solid_red" -> R.drawable.solid_red
            "solid_green" -> R.drawable.solid_green
            else -> R.drawable.gradient_blue
        }
        layout.setBackgroundResource(backgroundRes)
    }

    private fun applyPreferences() {
        val userId = getCurrentUserId()
        val theme = loadUserPreferences(userId)
        if (theme == "dark") {
            userListTextView.setBackgroundColor(getColor(android.R.color.black))
            userListTextView.setTextColor(getColor(android.R.color.white))
        } else {
            userListTextView.setBackgroundColor(getColor(android.R.color.white))
            userListTextView.setTextColor(getColor(android.R.color.black))
        }
    }

    private fun getCurrentUserId(): Int {
        return userManager.getUserInfo(android.os.Process.myUserHandle().identifier)?.id?.also {
            Log.d(TAG, "getCurrentUserId: ID: $it")
        } ?: run {
            Log.e(TAG, "getCurrentUserId: UserInfo is null")
            -1
        }
    }

    companion object {
        private const val TAG = "PersonalizedAndroid"
    }
}

